src/computability/regular_expressions.lean

```diff
@@ -0,0 +1,399 @@
+    { right,
+      exact ⟨ y, hy, z, hz, hx ⟩ } },
+  { rintro (⟨ y, hy, z, hz, hx ⟩ | ⟨ y, hy, z, hz, hx ⟩);
+    refine ⟨ y, z, _, hx ⟩,
+    { left,
+      exact ⟨ hy, hz ⟩ },
+    { right,
+      exact ⟨ hy, hz ⟩ } }
+end
+
+private lemma right_distrib (l m n : language α) : (l + m) * n = (l * n) + (m * n) :=
+begin
+  ext x,
+  simp only [mul_def, set.mem_image, add_def, set.mem_prod, exists_and_distrib_left, set.mem_image2,
+    set.image_prod, set.mem_union_eq, set.prod_union, prod.exists],
+  split,
+  { rintro ⟨ y, (hy | hy), z, hz, hx ⟩,
+    { left,
+      exact ⟨ y, hy, z, hz, hx ⟩ },
+    { right,
+      exact ⟨ y, hy, z, hz, hx ⟩ } },
+  { rintro (⟨ y, hy, z, hz, hx ⟩ | ⟨ y, hy, z, hz, hx ⟩);
+    refine ⟨ y, _, z, hz, hx ⟩,
+    { left,
+      exact hy },
+    { right,
+      exact hy } }
+end
+
+instance : semiring (language α) :=
+{ add := (+),
+  add_assoc := by simp [set.union_assoc],
+  zero := 0,
+  zero_add := by simp,
+  add_zero := by simp,
+  add_comm := by simp [set.union_comm],
+  mul := (*),
+  mul_assoc := mul_assoc,
+  zero_mul := by simp,
+  mul_zero := by simp,
+  one := 1,
+  one_mul := one_mul,
+  mul_one := mul_one,
+  left_distrib := left_distrib,
+  right_distrib := right_distrib }
+
+lemma add_self (l : language α) : l + l = l := by finish
+
+end language
+
+
+/-- This is the definition of regular expressions. The names used here is to mirror the definition
+  of a Kleene algebra (https://en.wikipedia.org/wiki/Kleene_algebra).
+  `zero` matches nothing
+  `epsilon` matches only the empty string
+  `char a` matches only the string 'a'
+  `star P` matches any finite concatenation of strings which match `P`
+  `plus P Q` matches anything which match `P` or `Q`
+  `comp P Q` matches `x ++ y` if `x` matches `P` and `y` matches `Q` -/
+inductive regular_expression (α : Type u) : Type u
+| zero : regular_expression
+| epsilon : regular_expression
+| char : α → regular_expression
+| plus : regular_expression → regular_expression → regular_expression
+| comp : regular_expression → regular_expression → regular_expression
+| star : regular_expression → regular_expression
+
+namespace regular_expression
+
+instance : inhabited (regular_expression α) := ⟨zero⟩
+
+instance : has_add (regular_expression α) := ⟨plus⟩
+instance : has_mul (regular_expression α) := ⟨comp⟩
+instance : has_one (regular_expression α) := ⟨epsilon⟩
+instance : has_zero (regular_expression α) := ⟨zero⟩
+
+@[simp] lemma zero_def : (zero : regular_expression α) = 0 := rfl
+@[simp] lemma one_def : (epsilon : regular_expression α) = 1 := rfl
+
+@[simp] lemma plus_def (P Q : regular_expression α) : plus P Q = P + Q := rfl
+@[simp] lemma comp_def (P Q : regular_expression α) : comp P Q = P * Q := rfl
+
+/-- `matches P` provides a language which contains all strings that `P` matches -/
+def matches : regular_expression α → language α
+| 0 := 0
+| 1 := 1
+| (char a) := {[a]}
+| (P + Q) := P.matches + Q.matches
+| (comp P Q) := P.matches * Q.matches
```

Comment:
`*` doesn't work here because `has_mul.mul` only has a `@[pattern]` annotation if you import `algebra.free`.

@gebner, do you happen to know where `has_add.add` gets its annotation from? `#print has_add.add` shows it, but not where it comes from.

Reply:
We can also add the attribute in mathlib.  I don't think it would cause any trouble to add it for `has_mul.mul`.

What `@[pattern]` does, is change the meaning of the name when used in a pattern:
```lean
match a with zero := zero end
-- zero is a variable

attribute [pattern] zero

match a with zero := zero end
-- zero is not a variable (and the match might be non-exhaustive)
```

Reply:
How should I resolve this? Would it make sense to import `algebra.free` or just add it to this file?

Reply:
Adding it to the file, perhaps with a TODO comment, is probably reasonable.